spec/enumerables_spec.rb

```diff
@@ -0,0 +1,273 @@
+    end
+  end
+
+  describe '#my_each_with_index' do
+    it 'should raise an error when an argument is given' do
+      expect { array1.my_each_with_index(1) }.to raise_error(ArgumentError)
+    end
+
+    context 'when block given' do
+      let(:hash) { {} }
+      it 'should call block with two arguments, the item and its index for each collection item' do
+        array2.my_each_with_index { |ele, idx| hash[ele] = idx }
+        expect(hash).to eq({ 'ten' => 2, 'three' => 1, 'twelve' => 3, 'two' => 0 })
+      end
+    end
+
+    context 'when no block is given' do
+      it 'should return an Enumerator' do
+        expect(array1.my_each_with_index).to be_a_kind_of(Enumerator)
+        expect(array1.my_each_with_index.to_a).to eq(array1)
+      end
+    end
+  end
+
+  describe '#my_select' do
+    it 'should raise an error when an argument is given' do
+      expect { array1.my_select(1) }.to raise_error(ArgumentError)
+    end
+
+    context 'when block given' do
+      it 'should return an array of all collection items that evaluate to true' do
+        expect(range.my_select { |ele| ele > 2 }).to eq([3, 4, 5])
+        expect(array2.my_select { |ele| ele.size > 3 }).to eq(%w[three twelve])
+      end
+    end
+
+    context 'when no block is given' do
+      it 'should return an Enumerator' do
+        expect(array1.my_select).to be_a_kind_of(Enumerator)
+        expect(array1.my_select.to_a).to eq(array1)
+      end
+    end
+  end
+
+  describe '#my_all?' do
+    context 'when no block is given' do
+      context 'if none of elements is (false) or (nil)' do
+        it 'should return true' do
+          expect(array2.my_all?).to be true
+        end
+      end
+
+      context 'if at least one or more elements is (false) or (nil)' do
+        it 'should return false' do
+          expect(array3.my_all?).to be false
+        end
+      end
+    end
+
+    context 'when block given' do
+      it 'should return true if all of the collection items evaluate to true' do
+        expect(array5.my_all? { |ele| ele >= 1 }).to be true
+        expect(range.my_all? { |ele| ele.is_a?(Numeric) }).to be true
+      end
+
+      it 'should return false if any of the collection items evaluates to false' do
+        expect(array3.my_all? { |ele| ele.is_a?(Integer) }).to be false
+        expect(range.my_all? { |ele| ele < 3 }).to be false
+      end
+    end
+
+    context 'when pattern is given as argument' do
+      it 'should return true if pattern === each element in collection' do
+        expect(array1.my_all?(Integer)).to be true
+        expect(array2.my_all?(/t/)).to be true
+      end
+
+      it 'should return false if pattern doesn\'t match each element in collection' do
+        expect(array3.my_all?(Integer)).to be false
+        expect(array1.my_all?(/e/)).to be false
+      end
+    end
+  end
+
+  describe '#my_any?' do
+    context 'when no block is given' do
+      context 'if  atleast one of elements is not (false) or (nil)' do
```

Comment:
**Optional**: try to keep your nesting simpler. If you write:
```
describe '#my_any?' do
    context 'when no block is given' do
        it 'should return true 'if at least one of elements is not (false) or (nil)'' do
          expect(array3.my_any?).to be true
        end
     end
// other context...
end
```
The code is more readable that way, and you only keep two contexts: `when block is given` and `when no block is given`.

Keep this in mind for all the other tests, if you decide to implement the change.